### ISSUE TITLE
CDAUDIO automount

### ISSUE THREAD

**Issue by Akuma-Git** [dosbox-staging/dosbox-staging#596]
**Is your feature request related to a problem? Please describe.**
Ex. I'm always frustrated when [...]
I need to create mundane CUE sheets when this can be automated
(cue sheet example below)

**Describe the solution you'd like**
After starting DosBox, check if files in folder CDAUDIO, and mount those tracks as first available drive letter D,E,F...etc.

**Describe alternatives you've considered**
Manual labor, hated it.

**Additional context**
CUE sheet example:

FILE "TRACK02.OPUS" MP3
  TRACK 01 AUDIO
    INDEX 01 00:00:00
FILE "TRACK02.OPUS" MP3
  TRACK 02 AUDIO
    INDEX 01 00:00:00
FILE "TRACK03.OPUS" MP3
  TRACK 03 AUDIO
    INDEX 01 00:00:00
FILE "TRACK04.OPUS" MP3
  TRACK 04 AUDIO
    INDEX 01 00:00:00
FILE "TRACK05.OPUS" MP3
  TRACK 05 AUDIO
    INDEX 01 00:00:00
FILE "TRACK06.OPUS" MP3
  TRACK 06 AUDIO
    INDEX 01 00:00:00
FILE "TRACK07.OPUS" MP3
 TRACK 07 AUDIO
    INDEX 01 00:00:00
FILE "TRACK08.OPUS" MP3
  TRACK 08 AUDIO
    INDEX 01 00:00:00

**Comment by dreamer** [comment]
@Akuma-Git where do you suggest the CDAUDIO directory is supposed to be located? How is dosbox-staging supposed to correlate CDAUDIO directories to specific games? Also a side note: your example has .opus files listed but instructs decoder to use mp3 codec, which is incorrect; I believe we deal with such incorrect .cue files automatically, but I sure hope we're generating warnings in the process...

@kcgen Can we even generate workable .cue file automatically based on the audio files alone? What about index values? There's no way for us to guess them correctly, I think… (but maybe a majority of games use `index 1` or something.

**Comment by kcgen** [comment]
Ideally we wouldn't have to make assumptions about a directory name.
The user could do something like `imgmount d mycdrom -t cdrom`, and then:
 - If only an ISO was found in `mycdrom/`, then mount it to D:\
 - If an ISO and one or more audio files where found in `mycdrom/` then mount the ISO and audio files in-name-sorted order.
    1. Assume a 2-second pregap for track 2, and then no pregap for the remainder. 
    1. We could also take the pregaps in as an argument. `-first-pregap 02:00` `-next-pregaps 02:00`.
    1. This would cover all the CDDA variations I've ever seen:
        - 99% just a have a 2-second pregap on track 2, and none on all the rest
        - Some have a unique track 2 pregap and then a different pregap for all the rest. 
        - And some have the same pregap across-the-board (in which case, set both arguments to the same value).

> lso a side note: your example has .opus files listed but instructs decoder to use mp3 codec, which is incorrect

Yes, the "TYPE" field should also be OPUS.  Fortunately the code inspects the actual file extension first and tries the corresponding decoder on it.  If that fails, then it tries the the provided TYPE field (in this case, MP3), and finally if that fails, it tries all the decoders until one succeed. . I believe warnings are produced when it's trying and failing - but then an error is produced when it's out of options.

**Comment by Akuma-Git** [comment]
> @Akuma-Git where do you suggest the CDAUDIO directory is supposed to be located? How is dosbox-staging supposed to correlate CDAUDIO directories to specific games? Also a side note: your example has .opus files listed but instructs decoder to use mp3 codec, which is incorrect; I believe we deal with such incorrect .cue files automatically, but I sure hope we're generating warnings in the process...

Sorry about that, I get sloppy when its the Xth game I do. Do you want me to edit it ?
The root folder where DOSBox starts looks like the best option now.

Those pregaps became obsolete according to: https://en.wikipedia.org/wiki/Pregap

**Comment by kcgen** [comment]
Unfortunately if we didn't handle pregaps, some games would ask for playback too far back or ahead; likewise their seeks would also be off.  Many games hardcode the actual CDROM sector of variable chunks of audio, and that assumes it will be offset by the pregap. (I would love to be able to eliminate all that from the code if we could!)

**Comment by dreamer** [comment]
> Sorry about that, I get sloppy when its the Xth game I do. Do you want me to edit it ?

No, no :) I just want us to be on the same page when designing the solution that everyone will be comfortable with :)

> The root folder where DOSBox starts looks like the best option now.

Hmm, there is no such root dir; dosbox looks for relative and absolute paths, like any other program.

> Those pregaps became obsolete according to: https://en.wikipedia.org/wiki/Pregap

Article strictly refers to hiding binary data in pregap as an obsolete technique. Pregap is mostly used to define the amount of silence between the audio tracks - some albums have no pregaps at all (as to provide continuous audio throughout all the tracks), while others have a small amount of silence between the tracks (usually 2s).

**Comment by dreamer** [comment]
@kcgen Instead of extending `imgmount` we should consider inventing a prefix scheme - similar to how Wine handles things? e.g. if we followed Wine, then we could have something like this:
```
game/drive_a_img/      # <- directory containing files to be imgmounted
game/drive_c/          # <- directory to be mounted as C:
game/drive_d_img/      # <- directory containing files / directories to be imgmounted
game/drive_d_img/cd1/  # <- 1st CD
game/drive_d_img/cd2/  # <- 2nd CD
game/name_slug.conf    # <- game-specific conf file
```
(in other words: directories `drive_x` to be mounted as X:, directories `drive_x_img` to be auto-detected and handled with imgmount)

And if with such standard layout, user could invoke the game like this: `dosbox --run-prefix game`

**Comment by Akuma-Git** [comment]
Ohh, I see where you are getting at, people have a whole structures containing multiple games. I usually just start DOSBox with the folder where my game resides eg:
/home/user/dosbox/mygame becomes c:\ to keep it isolated.

**Comment by dreamer** [comment]
@Akuma-Git yeah, we see all kind of schemes - some people have games isolated, some people do not (both solutions have their pros and cons). Personally I prefer to have games separated like you do… unless we're talking about testing - with having several games in one dir it's a bit easier to share `ultradir` or have Norton Commander always available in `PATH`.

**Comment by Akuma-Git** [comment]
I handle that by having a tools dir mounted at Y:, gus works fine with y:\ultradir.
[autoexec]
mount y "~/.dosbox/drive_y"
call y:\autoexec.bat
